### ISSUE TITLE
Video element breaks widget functionality in Firefox

### ISSUE THREAD

**Issue by garyapps** [ckeditor/ckeditor5#7525]
## 📝 Provide detailed reproduction steps (if any)

1. Create a plugin that inserts a `<figure>` element which has a `<video>` element inside it. The  `<video>` element should have a 'controls' attribute, which tells the browser to show controls such as a play button. 

2. As part of the code, make sure that the inserted video element is a widget so that it can be selected. (In the elementToElement downcast you would have a line that is something like this
`return toWidget( figure, viewWriter, { label: 'Video widget' } );` 
to make it a widget) 

3. Test on Chrome and Firefox. 

a) On Chrome everything works fine. If you hover over the video the yellow outline will appear, indicating it can be selected. If you click the video the outline appears blue, indicating it is selected. If you then click on some other widget in the editor the blue outline disappears from the video and appears on whatever other widget you clicked. In other words it works perfectly.

b) On Firefox however there is a problem. If you hover over the video the yellow outline will appear. But if you click the video the outline remains yellow. Instead the video starts playing. It is not a problem that the video starts playing, the problem is the yellow outline remains yellow, meaning it is impossible to select a video (and therefore also impossible to delete it). Moreover once this happens, it becomes impossible to select any other widget in the editor as well, whether video or something else. Basically the widget functionality breaks down completely after this point.

### ✔️ Expected result

Hovering over the video, which is a widget, should show the yellow outline, and clicking it should show the blue outline (indicating it is selected).

### ❌ Actual result

On Firefox, hovering over the video shows the yellow outline, but it remains yellow even after clicking. It is impossible to select it even though it is a widget. (The video starts playing, but the widget does not get selected). And at this point it also becomes impossible to select any other widget, such as images and tables. 

## 📃 Other details

1. Important - If you remove the 'controls' attribute from the video then in Firefox everything works perfectly. (The video does not start playing upon clicking it, and the widget gets selected successfully and shows a blue outline).

So my suspicion is that the built-in Firefox functionality, due to which a video plays upon clicking anywhere inside it, is interfering with the widget functionality. By removing the 'controls' attribute we are disabling that Firefox functionality, which seems to fix the problem. Unfortunately I need to have controls enabled in my app.

(Note that on Chrome the video does not play just by clicking anywhere on it, you have to click the play button on the video, but after you do that the widget functionality also continues to work perfectly. So the playing of the video is not a problem on Chrome, just on Firefox. Maybe the difference is in how the two browsers are capturing the click event). 

2. Initially I did not have a `<figure>` tag, I had a standalone `<video>`  tag. The problem was the same. I was hoping that by wrapping it in a  `<figure>`  it would fix t he problem but it did not.
 

* Browser: Firefox
* OS: Win 10
* CKEditor version: 20 
* Installed CKEditor plugins: -

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by Mgsy** [comment]
Hi, thanks for the report. It's hard to tell why this behaviour occur without knowing the whole implementation, however, I suggest checking our [implementing a block widget guide](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/implementing-a-block-widget.html) to get a basic concepts and analyze [Image plugin](https://github.com/ckeditor/ckeditor5/tree/65a9fc7049eee703ad02a9aca9620b8b1be0fa3e/packages/ckeditor5-image/src) as a reference.

**Comment by garyapps** [comment]
@Mgsy I have been through both the references you mentioned. In fact that is what I have been using to learn how to make widgets :) 

Since the widget works perfectly in Chrome, it seems to indicate that my code is fine, the problem is only on Firefox (and that too only when the 'controls' attribute is present on the video tag due to which the video automatically plays whenever it is clicked in Firefox). 

I also recently tested it on Edge and the code works fine on that browser too (Although I suppose that is to be expected since Edge uses the same rendering engine as Chrome).

**Comment by Mgsy** [comment]
Thanks for the details! Could you please prepare some working example where we'll be able to reproduce the issue?

**Comment by garyapps** [comment]
I will try to put a simple example together and get back to you.

**Comment by garyapps** [comment]
I have attached a sample. It is the basic inline plugin (using stable branch of the latest version from github repo) with all the default widgets, plus my simple video widget. The source files for the video widget are at the parent level. Note that I have updated the webpack.config.js file so it can find/resolve the location of these files during the build process. And I have updated ckeditor.js to include the video plugin.  

To test it out
1. Unzip folder
2. Open a command prompt in its location
3. Run  `npm install` to install the necessary packages
4. Run `npm build` to build 
5. Use the html file Test File.html

Test in both Firefox and Chrome. You will see the problem in Firefox whereas Chrome should work perfectly. 

[ckeditor5-build-inline.zip](https://github.com/ckeditor/ckeditor5/files/4869365/ckeditor5-build-inline.zip)

**Comment by garyapps** [comment]
I have attached a sample. It is the basic inline plugin (using stable branch of the latest version from github repo) with all the default widgets, plus my simple video widget. 

- The source files for the video widget are at the parent level. 
- Note that I have updated the webpack.config.js file so it can find/resolve the location of these files during the build process. 
- And I have updated ckeditor.js to include the video plugin.  

To test it out
1. Unzip folder
2. Open a command prompt in its location
3. Run  `npm install` to install the necessary packages
4. Run `npm build` to build 
5. Use the html file Test File.html

Test in both Firefox and Chrome. You will see the problem in Firefox whereas Chrome should work perfectly. 

[ckeditor5-build-inline.zip](https://github.com/ckeditor/ckeditor5/files/4869412/ckeditor5-build-inline.zip)

**Comment by Reinmar** [comment]
Let's see if I understand correctly:

*   clicking on a video in Chrome **selects** the widget _**and does not**_ **play** the video,
*   clicking on a video in Firefox **plays** the video and has some other negative side effects,
*   you could remove controls from the video (which fixes the above issue), but you actually want the video to be playable,
*   that also means that you cannot use a "mask" over the entire video (an element covering it so to prevent any events getting to the video).

Assuming that this is correct:

*   I guess that there may be something wrong with events being swallowed in Firefox and perhaps focus being lost. Unfortunately, such issues are usually hard to fix. Especially when they are browser specific. You could try a hack like https://github.com/ckeditor/ckeditor5/issues/4600#issuecomment-648814894 + using the `data-cke-ignore-events` on the `<video>` element or its wrapper. This may normalize the situation a bit (please let us know if it changes something). However, this will still not make the video selectable like other widgets. The problem is – its entire area plays the video instead of selecting the widget.
*   Therefore, you need to make some parts of the widget clickable:
    *   You can pass `options.hasSelectionHandle` to `toWidget()`. We use that in the table feature for the same reason – because the table area is to be used to interact with the table's contents, not to select the table.
    *   Cover part of the vide of a mask, leaving only the controls clickable. The mask would catch the interactions but you'd need to wisely position it so the controls are still clickable.
    *   Removing controls, rendering mask over the entire widget and adding a special button(s) e.g. in top right corner for playing the video. This is – rendering your own controls in such a place that will not affect the "selectability" of the widget too much.

I hope that helps :)

**Comment by garyapps** [comment]
Thanks @Reinmar . To give you an update on how I solved the issue, and for the benefit of anyone else reading -  

1. I now generate a UIElement as part of the editingDowncast - the way I understand it the contents of UIElements are ignored by CKEditor. 
2. I put the video tag inside of this UIElement
3. Then I run Video.js (which is one of the popular javascript video players) on this video element. Video.js generates its own UI and controls, so basically it replaces the browser controls with its own buttons and does its own event handling.

After these changes the video does not cause any issues for CKEditor in Firefox (and it still works correctly in Chrome). 

I assume the reason it works is because Firefox's built in video player was intercepting the clicks in certain browser-specific way that causes problems for CKEditor, whereas Video.js detects clicks in traditional javascript implementation, which is also consistent across browsers. Essentially it is doing what you suggested - rendering our own controls. Probably any other javascript video player (such as JWPlayer, jPlayer, FlowPlayer, etc) will also work equally well, I just happened to be familiar with this library.

[The reason I used the UIElement was because Video.js replaces the video tag with its own HTML, and so the page DOM will no longer match the model structure. By putting it in a UIElement I do not need to concern myself with the implementation details of Video.js and the structure of the HTML it generates (which is quite complex). Also this way what gets saved to my database is the single video tag rather than the html generated by Video.js].
